Sends video and logs to some external system for storage, indexing and reporting.

adrichem opened the pull request:
### Description
As discussed in https://github.com/zalando/zalenium/issues/430. I'm not sure how your documentation works. So I didn't update that. I hope you will do that. It needs to show, that
1. Setting an environment variable tells Zalenium where to send the artifacts to the store. My compose file uses this:
``` - REMOTE_DASHBOARD_HOST=http://artifact-store:4000 ```
1. If that environment variable does not exist, then the local dashboard will be used and Zalenium will not send the artifacts to the store.
1. The tester can tell Zalenium to send arbitrary JSON metadata together with the files. Here's a snippet of C# code that does that:
```
this.Driver.Manage().Cookies.AddCookie(new Cookie("zaleniumMetadata", JsonConvert.SerializeObject(Metadata)))
```

### Motivation and Context
See https://github.com/zalando/zalenium/issues/430. 

### How Has This Been Tested?
Been running it in our test environment for a few months. Today's commit implements an XOR choice between local or remote dashboard. That was tested by
1.  starting it with only local dashboard, running a session and seeing the video appear on the local dashboard.
2. starting it with remote dashboard, running a session and seeing that the dashboard remains empty.

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### Checklist:
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.
<!--- Provide a general summary of your changes in the Title above -->

diemol commented:
Thanks for this @adrichem!

Is it possible for you to rebase first? I know you mentioned in the issue that you were short of time but if you can, that would be really great. Afterwards, based on your instructions, I can update the docs.

adrichem commented:
You're welcome of course. 

I've rebased my branch onto your master. Can you please take a good look at the final set of diffs? I had a lot of conflicts due to age of the branch and the rename /replace of Dashboard.java to LocalDashboard.java .

Also please let me know if I need to tackle the Codacy\PR Quality Review issues,.

Cheers!

pearj commented on src/main/java/de/zalando/ep/zalenium/dashboard/remote/RemoteDriverLogDashboard.java:
```diff
@@ -0,0 +1,44 @@
+
+        List<FormField> fields = new ArrayList<>();
+
+        fields.add(new FormFile() {{
```
I don't think we want the "double-brace initialisation" anywhere, as it will probably create a garbage collection overhead, as well as potentially leaking memory.

Since this called after every test, it has more chance of being a problem.

I didn't even know that this "double-brace initialisation" was even a thing, until codacy picked it up: https://app.codacy.com/app/diemol_zalenium/zalenium/pullRequest?bid=4117327&prid=1752756#newIssuesView

I did a little digging and came across this article explaining why it's not a great idea: https://blog.jooq.org/2014/12/08/dont-be-clever-the-double-curly-braces-anti-pattern/

pearj commented on src/test/java/de/zalando/ep/zalenium/dashboard/LocalDashboardTest.java:
```diff
@@ -36,54 +36,76 @@
+            dashboard.updateDashboard(ti);
+            Assert.assertEquals(2, LocalDashboard.getExecutedTests());
+            Assert.assertEquals(2, LocalDashboard.getExecutedTestsWithVideo());
+        } catch (Exception e) {
```
Why do you catch the exception and rethrow it? I don't think you need to catch the exception at all. All it means is that the test errors will be harder to find the actual cause because the original exception will be lost.

adrichem commented:
Alright. I fixed almost all the Codacy issues and the unnecessary try/catch on the tests for the local dashboad (Doh!) One issue I did not fix was the warning about one unittest not containing an assert() or fail(). That's because the  .thenThrow() causes the assert() behavior. I'm not familiar enough with mockito to see how I could improve that.

 I also:
1. Reverted the rename and replace of Dashboard.java / LocalDashboard.java. It was  way to hard to check that I didn't accidentally lose any changes w.r.t. your master. Squashed those commits  i my branch so that we don't keep running into that again and again
1. Fixed a few compiler warnings introduced in the PR
1. Hit the 'Update branch' button in the PR to merge your latest commits from master

 Should  be good to go  now. If not, feedback appreciated!

diemol commented:
Thanks @adrichem for fixing the Codacy issues, I will merge this to a protected branch so Travis can run all the tests. After that, add some docs based on your comments and release. Hopefully it won't take so long.

Idsbali commented:
Can we also send other files (files getting downloaded in the browser) to some external system storage (like S3) ?
